Introduce reusable GCard component for unified card layout

itisAliRH opened the pull request:
This PR introduces a new reusable component called `GCard`. It's built to standardize the card layout used in different parts of the Galaxy UI and remove duplicate code across the project.
Until now, different parts of the UI (like workflows, notifications, etc.) have been using similar "card" layouts, but each has repeated and some custom styles.

## GCard Diagram
![GCard diagram](https://github.com/user-attachments/assets/9e85f008-8fd1-43af-8c5d-0fa200894173)



## GCard Highlights
- Easy to use by just passing props
- Responsive using CSS flex instead of grid
- Fully customizable — every section of the card can be overridden using slots.
- Can be used as a styled card wrapper for different use cases


## How to use it?
### Only Props
```vue
<GCard
  title="Card Title"
  description="This is an awesome component"
  :tags="['microbiome', '#Galaxy']"
  selectable
  :updateTime"updateTime"
  @select="onSelect"
/>
```

### Use slots
```vue
<GCard selectable :published="itemPublished">
  <template #title>
     <div class="my-component-title"> Custom Title </div>
  </template>

  <template #description> 
    <span class="shiny-description">    
      Some random shiny text 
    </span>
  </template>
</GCard>
```

### Use it as a layout wrapper only
```vue
<GCard>
  <div>Your custom content goes here</div>
</GCard>
```

## Updated Components

The following components have been updated to use the new `GCard` layout and more improvements:

### WorkflowCard
- Now uses the new GCard structure
- Unified look with improved layout and spacing
### Text summary 
- Long text descriptions now fully utilize available width.
- No awkward line breaks or clipping
### Tags
- Introduced an inline tag display mode
- Removes extra white space and fits better inside the card layout

|Before|
|---|
|![image](https://github.com/user-attachments/assets/e4868e59-d3cd-43cf-9c62-77d38651f263)|

|After|
|---|
|![image](https://github.com/user-attachments/assets/5973f94e-75f5-4459-8410-aeab081b1e1d)|

### Invocations list and card highlight

|Before|After|
|---|---|
|![image](https://github.com/user-attachments/assets/3aea00a0-6375-40e1-9006-73422570fc7e)|![image](https://github.com/user-attachments/assets/1551c42d-0165-43d2-a593-382005a7021f)|

### Notifications
|Before|
|---|
|![image](https://github.com/user-attachments/assets/0ba46d93-ff6b-475f-b4af-2b883466372e)|

|After|
|---|
|![image](https://github.com/user-attachments/assets/e6991cba-ac7e-45d3-8315-be15c9b1de67)|

### Archived Histories List

|Before|
|---|
|![image](https://github.com/user-attachments/assets/0bdbdb60-d480-4136-ba85-249da36393e8)|

|After|
|---|
|![image](https://github.com/user-attachments/assets/1d12d87d-91e1-498d-8929-d1074e11dbf5)|

### Storage locations templates

|Before|
|---|
|![image](https://github.com/user-attachments/assets/278b4f2f-26d3-45f6-803d-1920aa477f9d)|

|After|
|---|
|![image](https://github.com/user-attachments/assets/c456df4c-ed40-4781-ad51-fa8ab2529341)|

### Remote filles sources templates

|Before|
|---|
|![image](https://github.com/user-attachments/assets/4333e838-7815-40bc-9f45-61a7fec97267)|

|After|
|---|
|![image](https://github.com/user-attachments/assets/c34df688-53d4-4f9d-b92e-07849aeb55cf)|

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

itisAliRH committed:
Adds generic GCard component and type definitions

Introduces the GCard component with comprehensive props for customization.
Defines associated types for card attributes and badges.
Includes support for various actions, indicators, and UI elements.

itisAliRH committed:
Refactors ArchivedHistoryCard component

Replaces BootstrapVue components with GCard component
Updates imports and refactors badges and actions handling
Simplifies HistoryArchive component by removing BListGroup

Improves readability and maintainability and consistency

itisAliRH committed:
Refactors WorkflowCard component with GCard

Modularizes logic into composable functions
Replaces multiple imports with a single GCard component
Simplifies template structure and improves code readability and consistency
Enhances maintainability by separating concerns

itisAliRH committed:
Removes unused WorkflowActions components

Deletes WorkflowActions.vue, WorkflowActionsExtend.vue, and useWorkflowActions.ts to clean up unused code and reduce codebase complexity.

itisAliRH committed:
Replaces FontAwesome link icon with ExternalLink component

Updates DOI link display to use ExternalLink component for better reusability and consistency.
Removes FontAwesome icon dependency and associated styling.

itisAliRH committed:
Update Heading component icon prop type

Changes icon prop type from string to IconDefinition for better type safety.
Sets default icon prop value to undefined.

itisAliRH committed:
Refactors TextSummary component layout and logic

Refactor component to trim the text based on width of the component (not number of the words)
Replaces the component prop default value with "span"
Removes unnecessary HTML rendering logic
Simplifies text summary display handling
Updates styles for text overflow and ellipsis

itisAliRH committed:
Adds icon to 'never run' badge in WorkflowInvocationsCount

Includes FontAwesome icon to visually indicate workflows that have never been run, improving user interface clarity.

itisAliRH committed:
Enhances GCard component with new props and layout improvements

Adds `titleIcon` and `titleSize` props for customizable title display
Reorganizes header and footer layout for better responsiveness
Removes redundant title prop initialization
Updates styling for improved visual consistency

Fixes layout issues on smaller screens

itisAliRH committed:
Adds title badges to WorkflowCard component

Introduces workflowCardTitleBadges to display badges in the title of WorkflowCard.
Separates title-specific badges from general workflow badges for better clarity.

itisAliRH committed:
Simplifies WorkflowCardList.vue styling

Refactors WorkflowCardList.vue to use Bootstrap utility classes for layout and overflow management, removing custom CSS definitions.

itisAliRH committed:
Refactors InvocationScrollList component

Replaces BListGroupItem with GCard for better UI consistency
Adds new icons and badges for enhanced visualization
Removes unused UtcDate component

Improves code readability and maintainability

itisAliRH committed:
Improves GCard component layout

Aligns bookmark actions to the end in GCard component
Removes unnecessary conditional rendering of tags section
Fixes grid column alignment for small screens

itisAliRH committed:
Refactors SourceOptionCard to use GCard component

Replaces Heading and TextSummary with GCard
Simplifies template structure and styling
Enhances code readability and maintainability

itisAliRH committed:
Enhances tag display with tooltip and toggle button

Adds BTooltip for better tag preview
Introduces toggle button for displaying additional tags
Refactors tag iteration logic for improved readability

itisAliRH committed:
Adds inline property to CardAttributes and TitleIcon interface

Introduces an optional inline property to the CardAttributes interface
and creates a new TitleIcon interface to enhance flexibility in defining card icons.

itisAliRH committed:
Enhances GCard component with new icon support and UI improvements

Adds `updateTimeIcon` prop to display an icon before the update time
Changes `titleIcon` type to `TitleIcon` for better type safety
Sets default `titleSize` to `sm`
Refactors card header and layout for improved readability and maintainability
Updates CSS for better alignment and layout consistency

itisAliRH committed:
Fixes title-icon binding issue

Updates title-icon binding to use an object format for compatibility with the latest icon component API.

itisAliRH committed:
Fixes self-closing tag syntax for WorkflowCard

itisAliRH committed:
Refactors NotificationsPanel component

Removes unused FontAwesome library import and related initialization
Simplifies v-for loop in template
Eliminates unnecessary CSS class for notification cards

itisAliRH committed:
Refactors NotificationCard component

Replaces individual notification category components with a unified GCard component.
Adds computed properties for notification details.
Introduces primary action buttons for notifications.
Enhances UX with markdown rendering and dynamic icons.

Improves code maintainability and user experience.

itisAliRH committed:
Adds flex-gap utility classes to GCard component

Updates class definitions to include new flex-gap utility classes
for consistent spacing. Removes redundant gap properties from
CSS grid layouts for cleaner style management.

itisAliRH committed:
Removes redundant CSS positioning

Simplifies the extra actions div by removing unnecessary absolute positioning style.

itisAliRH committed:
Refactors GCard component layout

Updates GCard component to use more flexible and modern CSS classes for improved responsiveness and maintainability. Simplifies the structure by removing unnecessary grid-based styles and aligning elements using flexbox.

itisAliRH committed:
Refactors workflow card action handlers

Removes unnecessary arrow functions in handler assignments.

itisAliRH committed:
Update GCard element ID generation logic

Modifies ID generation functions to adjust element order for better readability and consistency for tests.
Updates bookmark button ID based on bookmark status.

itisAliRH committed:
Updates workflow components to use WorkflowSummary

Replaces StoredWorkflowDetailed with WorkflowSummary in workflow-related components to improve performance and streamline data handling.

itisAliRH committed:
Adds missing actions to workflow card

Introduces actions for viewing external links based on workflow source metadata
Adds logic to compute external URLs for Dockstore and other sources

itisAliRH committed:
Updates workflow selector identifiers

Revises CSS selectors for workflow card elements using GCard component

itisAliRH committed:
Simplifies workflow step addition methods

Replaces repeated workflow button search and click logic with direct method calls for adding steps and sub-workflows.

itisAliRH committed:
Updates class selector in NotificationsList tests

Changes the selector from ".notification-card" to ".g-card" to reflect updated class naming conventions in the NotificationsList component tests.

itisAliRH committed:
Updates notification preview component

Replaces MessageNotification with GCard for notification preview in NotificationForm.

davelopez reviewed (commented):
Very cool so far!

Some small issues I found while playing with this maybe some of those are already known:

- General all cards

  - [ ] "Last edited" tooltip displays double, one with time and one without, both with different styles. It should display only one including the full exact date.

- Workflow card

  - [x] (Known) The select box is not working properly
  - [ ] Card title should break on words
  - [ ] In bulk actions, "add tags" doesn't select tags properly
  - [x] "Imported From" button does not copy the link
  - [ ] Filter "Show bookmarked" raises a 500 error (backend issue?)

- Notification card

  - [x] Tooltip for "publication_time" could be "Sent on" instead of "Published on". It may be confusing since the notification is not "public"
  - [x] Marking as read does not update the card immediately

itisAliRH committed:
Updates Heading component icon prop type

Allows icon prop to accept either IconDefinition or string array to accept custom icons
Removes redundant SVG font-size styling

itisAliRH committed:
Enhances workflow WorkflowSummary

Adds optional source_type and detailed source_metadata fields to WorkflowSummary
Removes unnecessary TypeScript ignore comment and improves clarity of source type handling

itisAliRH committed:
Adds inline tags display and toggle functionality

Introduces inline prop to enable inline tags display
Enhances tag toggle button to show fewer tags with icon
Includes FontAwesome icons for improved UI/UX

itisAliRH committed:
Adds support for external links in GCard component

Introduces an `externalLink` property to enable opening links in a new tab
Updates card actions to utilize this property for external URLs
Enhances user experience by clearly differentiating external links

itisAliRH committed:
Refactors notification actions to use computed property

Changes primaryActions from a static list to a computed property for better reactivity.
Updates notification time title from 'Published' to 'Sent'.
Adds CSS styling to remove margins from paragraphs in notification messages.
Simplifies rendering of NotificationCard components in NotificationsList.

itisAliRH committed:
Corrects event emission order in WorkflowCard component

Reorders the emission of 'insertSteps' and 'insert' events to ensure proper sequence in WorkflowCard.vue.

itisAliRH committed:
Improves GCard.vue layout and accessibility

Refines CSS classes to enhance layout consistency
Adds margin and alignment adjustments for better UI

itisAliRH committed:
Adds margin reset to text elements

Ensures consistent spacing by resetting margin to 0 for text elements within TextSummary component.

itisAliRH committed:
Refines workflow and history panel selectors, improves tests

Updates CSS selectors for workflow and history panel elements to enhance specificity and maintainability.
Adds 'show more tags' button handling in history panel tests. Refines workflow management tests to utilize updated selectors.

itisAliRH committed:
Fix import paths for getFullAppUrl utility functions

itisAliRH committed:
Updates GCard component styling

itisAliRH committed:
Improve clickable functionality to GCard component

Introduces a 'clickable' prop to GCard to enable keyboard and click interactions.
Updates InvocationScrollList to utilize the new clickable prop for better accessibility.

itisAliRH committed:
Fixes selector and tooltip clearing in workflow navigation

Updates CSS selector for workflow invocations list items to match new class structure.
Adds tooltip clearing before workflow search to improve UI rendering.
Enhances tooltip clearing method to accept custom CSS selector.

itisAliRH committed:
Removes redundant class binding in InvocationScrollList

Simplifies the component template by removing an unnecessary `:class` binding for `panel-item` when rendering invocation items.

itisAliRH committed:
Refines InvocationScrollList selectors and icon

Replaces inline icon usage with a standalone FontAwesomeIcon component.
Updates CSS selectors for object store and file source creation actions to use consistent, descriptive naming.

itisAliRH committed:
Refactors notification card component

Consolidates multiple notification components into a shared `NotificationCard` component for improved maintainability and reusability.
Removes category-specific notification components and styles.
Updates tests to reflect the new component structure.

itisAliRH committed:
Removes NotificationActions component and consolidates logic

Deletes NotificationActions component and its tests, moving key functionalities to NotificationCard.

davelopez reviewed (approved):
Thank you!

I'm looking forward to using the new GCards!

ahmedhamidawan reviewed (approved):
Looks awesome, thanks!
